sourcegraph/sourcegraph.proto

```diff
@@ -588,6 +588,215 @@ service Repos {
 	rpc ListCommitters(ReposListCommittersOp) returns (CommitterList);
 }
 
+// StorageError represents an error when interacting with the Storage service.
+message StorageError {
+	// Code represents the type of error for programatic handling.
+	enum Code {
+		// NONE is the error code used when programatic handling of the error is not
+		// advised. It can be assumed that the operation failed, and that a human
+		// readable message is also provided.
+		NONE = 0;
+
+		// EOF is the error returned by Read when no more input is available.
+		// Functions should return EOF only to signal a graceful end of input. If
+		// the EOF occurs unexpectedly in a structured data stream, the appropriate
+		// error is either UNEXPECTED_EOF or some other error message giving more
+		// detail.
+		EOF = 1;
+
+		// NOT_EXIST is the error used when attempting to read/write to an object
+		// that does not exist.
+		NOT_EXIST = 2;
+
+		// PERMISSION is the error used when permission to access the given file is
+		// not granted to you.
+		PERMISSION = 3;
+	}
+
+	// code is the error code. If no error code is specified then programatic
+	// handling of the error is not advised. The user should be informed of the
+	// error message instead.
+	Code code = 1;
+
+	// message is the human-readable error message.
+	string message = 2;
+}
+
+// StorageName is a storage object's name.
+message StorageName {
+	// app_name is the name of the application whose data you are trying to
+	// read/write, applications may read and write to eachother's data assuming
+	// the admin has not restricted such access.
+	string app_name = 1;
+
+	// If specified storage is considered local to the given repository. Otherwise
+	// it is considered "global" (i.e. shared across all repositories).
+	string repo = 2;
+
+	// name is the name of the file.
+	string name = 3;
+}
+
+// StorageReadOp is the parameters for reading from a file.
+message StorageReadOp {
+	StorageName name = 1 [(gogoproto.nullable) = false];
+
+	// offset is the offset in bytes in which to perform the read operation from
+	// the start or end of the file, depending on offset_end. If no offset is
+	// specified, read operations always occur at the start of the file (they do
+	// not retain state).
+	int64 offset = 2;
+
+	// offset_end causes the offset to act relative to the end of the file, if
+	// set (i.e. offset == -100 would mean to read starting 100 bytes from the end
+	// of the file).
+	bool offset_end = 3;
```

Comment:
The rationale here is that this should be as-close-as-possible to the Go FS API as possible. In fact I would like to have a `whence` here instead, but because it is stateful that would put us into a bad position from a resource management perspective.

By keeping this stateless we could potentially close the file directly after the read operation, etc, which keeps open FDs lower. We will probably use a heuristic for this in the future. In general we can't fully trust gRPC clients to close the file, and this could be an attack vector: open as many files as possible and don't close them. The goal is to provide the best API possible (least user hassle) while still keeping it stateless.

Closing files from a user-of-the-API perspective should be considered mandatory. From a server implementation perspective, we should expect clients may never close files at all (to avoid this potential attack vector).

Also, it simplifies code for users who might be using this _not from Go_, I personally think.

I'll implement your other comments later today. I agree with them.